inst/include/dust/rng.hpp

```diff
@@ -12,67 +12,82 @@ namespace dust {
 
 // This is just a container class for state
 template <typename T>
-class pRNG { // # nocov
+class pRNG {
 public:
-  pRNG(const size_t n, const std::vector<uint64_t>& seed) {
-    rng_state_t<T> s;
+  pRNG(const size_t n, const std::vector<uint64_t>& seed) :
+    n_(n), state_(n * rng_state_t<T>::size()) {
     auto len = rng_state_t<T>::size();
     auto n_seed = seed.size() / len;
     for (size_t i = 0; i < n; ++i) {
+      rng_state_t<T> s = state(i);
       if (i < n_seed) {
-        std::copy_n(seed.begin() + i * len, len, s.state.begin());
+        for (size_t j = 0; j < len; ++j) {
+          s[j] = seed[i * len + j];
+        }
       } else {
+        rng_state_t<T> prev = state(i - 1);
+        for (size_t j = 0; j < len; ++j) {
+          s[j] = prev[j];
+        }
         xoshiro_jump(s);
       }
-      _state.push_back(s);
     }
   }
 
   size_t size() const {
-    return _state.size();
+    return n_;
   }
 
   void jump() {
-    for (size_t i = 0; i < _state.size(); ++i) {
-      xoshiro_jump(_state[i]);
+    for (size_t i = 0; i < n_; ++i) {
+      xoshiro_jump(state(i));
     }
   }
 
   void long_jump() {
-    for (size_t i = 0; i < _state.size(); ++i) {
-      xoshiro_long_jump(_state[i]);
+    for (size_t i = 0; i < n_; ++i) {
+      xoshiro_long_jump(state(i));
     }
   }
 
-  rng_state_t<T>& state(size_t i) {
-    return _state[i];
+  // Access an individual rng_state by constructing the small struct
+  // that contains a pointer to the memory, offset as needed, and our
+  // stride.
+  rng_state_t<T> state(size_t i) {
+    return rng_state_t<T>(state_.data() + i, n_);
```

Comment:
Why doesn't this one have `i * n_` as below?

Reply:
If they are meant the be identical, could you use a define to set the functions as being the same? Or was that the fiddly bit you mentioned?

Reply:
The other one was vestigial - I've removed it and if the tests pass that confirms it...